File: src/main/js/formulas/logic/IF.ts
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class IF extends AbstractFormula {
  static SELF: IF = new IF();
  NAME = FormulaName.LT;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 3, this.NAME);
    const logicalExpression = Converters.toBoolean(
      Converters.first(this.collateralLookup(origin, values[0]))
    );
    if (logicalExpression === true) {
      return Converters.first(this.collateralLookup(origin, values[1]));
    }
    return Converters.first(this.collateralLookup(origin, values[2]));
  }
}

File: src/main/java/io/protobase/f7/formulas/logic/IF.java
package io.protobase.f7.formulas.logic;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class IF extends AbstractFormula {
  public static FormulaName NAME = FormulaName.IF;
  public static IF SELF = new IF();

  public IF() {
    super();
  }

  public IF(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 3, NAME);
    Boolean logicalExpression = Converters.toBoolean(Converters.first(collateralLookup.apply(origin, values[0])));
    if (logicalExpression) {
      return collateralLookup.apply(origin, values[1]);
    }
    return collateralLookup.apply(origin, values[2]);
  }
}
